src/main/scala/org/ergoplatform/mining/ErgoMiner.scala

```diff
@@ -268,35 +270,33 @@ class ErgoMiner(ergoSettings: ErgoSettings,
     }
   }
 
-  // We produce block candidate on getting readers for node's view from NodeViewHolder
-  private def onReaders: Receive = {
-    // Miner's node can produce block candidate only if it is working in the UTXO regime
-    case Readers(h, s: UtxoStateReader, m, _) =>
-     generateCandidate(h, m, s, None)
-  }
-
   private def mining: Receive = {
-    case PrepareCandidate(_) if !ergoSettings.nodeSettings.mining =>
+    case PrepareCandidate(_, _) if !ergoSettings.nodeSettings.mining =>
       sender() ! Future.failed(new Exception("Candidate creation is not supported when mining is disabled"))
 
-    // Send cached candidate if it is available and (non-empty) list of transactions to include hasn't been changed
-    case PrepareCandidate(txsToInclude) =>
-      if (cachedFor(txsToInclude)) {
-        val candBlockFuture = candidateOpt
-          .map(_.externalVersion)
-          .fold[Future[WorkMessage]](
-          Future.failed(new Exception("Failed to create candidate")))(Future.successful)
-        sender() ! candBlockFuture
+    case PrepareCandidate(txsToInclude, reply) =>
+      val candF: Future[CandidateCache] = if (candidateGenerating) {
+        Future.failed(new Exception("Skipping candidate generation, one is already in progress"))
       } else {
-        log.info("Generating new candidate requested by external miner")
-        val readersR = (readersHolderRef ? GetReaders).mapTo[Readers]
-        val candBlockFuture = readersR.flatMap {
-          case Readers(h, s: UtxoStateReader, m, _) =>
-            Future.fromTry(generateCandidate(h, m, s, Some(txsToInclude)).map(_.externalVersion))
-          case _ =>
-            Future.failed(new Exception("Invalid readers state, mining is possible in UTXO mode only"))
+        candidateGenerating = true
+        val f = if (cachedFor(txsToInclude)) {
+          candidateOpt.fold[Future[CandidateCache]](
+            Future.failed(new Exception("Failed to create candidate")))(Future.successful)
+        } else {
+          log.info("Generating new candidate requested by external miner")
+          val readersR = (readersHolderRef ? GetReaders).mapTo[Readers]
+          readersR.flatMap {
+            case Readers(h, s: UtxoStateReader, m, _) =>
+              Future.fromTry(generateCandidate(h, m, s, Some(txsToInclude)))
+            case _ =>
+              Future.failed(new Exception("Invalid readers state, mining is possible in UTXO mode only"))
+          }
         }
-        sender() ! candBlockFuture
+        f.onComplete(_ => candidateGenerating = false)
+        f
+      }
+      if (reply) {
+        sender() ! candF.map(_.externalVersion)
       }
```

Comment:
I don't like the idea of passing Future as a message in general. 
Here, more specifically, cannot imagine what the state it is when the sender() receives the Future which is still being processed in parallel somewhere in readersHolderRef and "generateCandidate" is not even started.
This part needs major rethinking, IMO.

Reply:
@aslesarenko that's typical for our API though